.github/workflows/github-actions-release.yml

```diff
@@ -98,22 +98,29 @@ jobs:
           SOURCE_SHA=$(git rev-parse HEAD);
           TAG_NAME="${{ steps.resolve-tag.outputs.tag_name }}"
 
-          cd ./packages/js/github-actions
+          pushd ./packages/js/github-actions
           npm ci --ignore-scripts
           npm run build
-          cd -
+          
+          cd ../../php/github-actions
+          ./build.sh
+          popd
 
           git config user.name github-actions[bot]
           git config user.email 41898282+github-actions[bot]@users.noreply.github.com
 
           git rm -r .
           git commit -q -m "Create the ${TAG_NAME} release build for the \`github-actions\` package."
 
-          git checkout HEAD^ -- ./packages/js/github-actions/actions
+          git checkout HEAD^ -- ./packages/js/github-actions/actions ./packages/php/github-actions
           git restore --staged .
           echo "/packages/js/github-actions/actions/*/src" > .gitignore
           git add ./packages/js/github-actions/actions
           git mv ./packages/js/github-actions/actions/* ./
+          
+          find ./packages/php/github-actions -name "*.zip" -exec unzip -d ./ {} \;
+          git add .
```

Comment:
> Ideally, this process builds a release revision that can be run directly in node or php, with the required non-built-in dependencies also inlined.

While it's possible to consolidate down into a single file in PHP (or technically, an archive), I don't think the desire for a single file necessitates the overhead of building that. Generally speaking, PHP applications span multiple files, even in production.

> However, this made the [packages](https://github.com/woocommerce/grow/tree/a14aa245eb025c2d90a0f546ca2fd4bae6698c5a/packages) directory to be committed together, and the directory includes:
> 
> - The source codes of `php/github-actions`
> - The source codes of each action in `js/github-actions`
> - `hook-documentation.zip` file
> - `node_modules` directory
> - `vendor` directory
> 
> In addition, there are a few files that were committed as well, such as the `.gitignore` in the root directory. [The contents of this link](https://github.com/woocommerce/grow/tree/actions-v1.4.1?rgh-link-date=2023-05-12T08%3A42%3A27Z) can be a reference for the build results.
> 
> Could it still make it only commit the minimum required files?

Good catch. My goal was to build this in such a way to allow for future actions to be developed, and the release process handles them automatically. But as you point out, simply using `git add .` adds more than I intended.

(As an aside, I understand *why* we do the build this way, but I think it would be worth reconsidering how the repo is managed to facilitate a more intuitive process, and to also prevent warnings that the tag commits aren't part of the repo.)

I added 8d1189bcb09007fc8ffb1b68eea685e8f1904d9d to limit the files that are included during the commit process.

Reply:
> While it's possible to consolidate down into a single file in PHP (or technically, an archive), I don't think the desire for a single file necessitates the overhead of building that. Generally speaking, PHP applications span multiple files, even in production.

Thanks for the clarification.

Since the npm ecosystem can easily bloat the content in `node_modules`, it's not ideal to commit the directory together for a release. Running `npm install` on the client side before performing it may take longer and have compatibility issues. Therefore, the practice of GitHub Actions written in JavaScript is usually to build them into a single file or a handful of files, and there are opportunities to reduce the overall size through tree-shaking techniques.

I agree that the PHP case is different, and guess whether or not to do further build processing is more dependent on the actual use case rather than the way in JS.

> (As an aside, I understand _why_ we do the build this way, but I think it would be worth reconsidering how the repo is managed to facilitate a more intuitive process, and to also prevent warnings that the tag commits aren't part of the repo.)

The main reason for the intuitive build process is because of the above npm ecosystem and compatibility considerations.

Another consideration is to minimize the time spent on the private repos to use these GH Actions, which in turn reduces the cost spent on GH. Therefore, the size of the release and the direct executability are also the main indicators.

Not sure about the "tag commits aren't part of the repo" warnings. Could you provide me with a practical example?

Reply:
> Since the npm ecosystem can easily bloat the content in node_modules, it's not ideal to commit the directory together for a release. Running npm install on the client side before performing it may take longer and have compatibility issues. Therefore, the practice of GitHub Actions written in JavaScript is usually to build them into a single file or a handful of files, and there are opportunities to reduce the overall size through tree-shaking techniques.
> 
> I agree that the PHP case is different, and guess whether or not to do further build processing is more dependent on the actual use case rather than the way in JS.

Thanks for the explanation about the NPM process, it makes sense why you'd want to do all of the processing and consolidation ahead of time. 

With PHP's `composer` for dependency management, it is fairly efficient when set up correctly. Just like with NPM, you can define production vs. development dependencies, and composer allows for loading up dependencies that are only needed for production. In particular, the build process that's executed earlier builds a production-ready version of the dependencies, and utilizes an optimized autoloader for file inclusion.

In short, the building that's done to generate the zip files is the most efficient version of the project possible, and we're only including what's necessary for the project to run. Further, the testing I've done shows that it completes very quickly.

> Not sure about the "tag commits aren't part of the repo" warnings. Could you provide me with a practical example?

If you view the commit for any of the releases, or just browse the codebase from one of the release tags, GitHub shows a warning. Here's an example: https://github.com/woocommerce/grow/tree/actions-v1.4.1

Reply:
> > Not sure about the "tag commits aren't part of the repo" warnings. Could you provide me with a practical example?
> 
> If you view the commit for any of the releases, or just browse the codebase from one of the release tags, GitHub shows a warning. Here's an example: https://github.com/woocommerce/grow/tree/actions-v1.4.1

Since this repo also contains other packages and has to support multiple Custom GH actions in a single repo, (most of the Custom GH actions are one action for one repo construct), we came to the current structure after some discussion. Related discussions are in #12

To reduce confusion, the README.md below [actions-v1.4.1](https://github.com/woocommerce/grow/tree/actions-v1.4.1) briefly explains that it is a version-specific release build and guides to the actual revision of the source code. The development page also demonstrates the composition of version tagging in the git tree.

The reason for intentionally not having release commits and tags be created on a `trunk` or a specific branch is to avoid that branch's commit history looking more confusing. After all, each build removes all other packages and moves itself to the root directory, so if that branch has this particular commit and then reverts it, there may not be less complexity.

With these limitations and premises, it does result in some complexity and is not intuitive, and I would be happy to see if there is a better way to construct it and still fulfill the needs.

- Coexist with other packages in the grow repo
- Support multiple Custom GH actions in a single repo
- Released Custom GH actions are kept as thin and ready to run immediately as possible
- The caller repos can always use a path in `woocommerce/grow/<action-dir>@actions-v1` pattern